### ISSUE TITLE
Add option to specify paths for ScriptAnalyzer-evaluation

### ISSUE THREAD

**Issue by fflaten** [PowerShell/vscode-powershell#3048]
**Summary of the new feature**

This extension should include a setting to specify which paths to run Script Analysis against. Currently the extension reads the PSScriptAnalyzer.psd1-file in workspace root (if not modified) and executes PSSA against all files recursively.

Let's say you only want to run the rules against src/* and tool/*. AFAIK this can't be configured in the extension nor `PSScriptAnalyzer.psd1`, which can result in many unnecessary problems and warnings in folders without the same requirements to follow best practices.

**Proposed technical implementation details (optional)**

`Invoke-ScriptAnalyzer` solves this problem by requiring a Path or ScriptDefinition parameter. As the extension uses the latter option, then the logic and paths-setting would have to be part of the Script Analysis-part of vscode-powershell + PowerShellEditorServices.

**Comment by rjmholt** [comment]
So I think adding a configuration like this to PSScriptAnalyzer's settings file makes sense, and for that I'd ask you to open a feature request issue in the PSScriptAnalyzer repo.

On the VSCode extension front, I want to understand the scenario a bit better.

You've said in the issue description:

> Currently the extension reads the PSScriptAnalyzer.psd1-file in workspace root (if not modified) and executes PSSA against all files recursively.

Is that what you're observing?

Currently diagnostics are triggered only when [a file is opened](https://github.com/PowerShell/PowerShellEditorServices/blob/b243fe3dd3baf88399b370437a827778b86e24ad/src/PowerShellEditorServices/Services/TextDocument/Handlers/TextDocumentHandler.cs#L94) or [edited](https://github.com/PowerShell/PowerShellEditorServices/blob/master/src/PowerShellEditorServices/Services/TextDocument/Handlers/TextDocumentHandler.cs#L61), and analysis [is only run on those files](https://github.com/PowerShell/PowerShellEditorServices/blob/b243fe3dd3baf88399b370437a827778b86e24ad/src/PowerShellEditorServices/Services/Analysis/AnalysisService.cs#L383-L390).

So if you're seeing the VSCode extension do something different, that's probably a bug.

Given that that's the intended behaviour, a configuration to exclude files just from script analysis is something we wouldn't really look to add unless there's a particular scenario you're trying to address.

Could you elaborate a bit on your scenario, what you're seeing and what the current vs ideal behaviour is?

**Comment by fflaten** [comment]
Yes, it only analyzes open files - but it analyzes *any* open powershell-file in the workspace. There's currently no way to limit vscode-powershell/PSES analysis to specific folders.

I don't need the same quality checks in /tst/, /tools/ etc. as I would /src/ which is the acutal published code.

A setting could be included in PSScriptAnalyzer.psd1, but PSES uses `Invoke-ScriptAnalyzer -ScriptDefintion ...` so any path-check can't be processed by PSSA because the path for the currently opened file is never provided to themodule. So this exclude/include-setting would need to be processed in vscode-powershell/PSES AFAIK.

**Comment by simonsabin** [comment]
Would it be better to have the extension look for psd1 in each folder. That way you can specify different rules for different folders. That also then allows you to match the Invoke when you specify a path to analyse and the path to the psd1. 
Having it in settings feels a bit limiting

**Comment by fflaten** [comment]
That was probably gonna be my next feature request, by suggestion to extend the settings.json to support of multiple PSSA settings-files, each with their own path-settings.

Using a new psd1-file in each  folder could easily lead to duplicate files.

In this include/exclude folder-scenario you'd also have to create a new psd1 in every folder where PSES shouldn't analyze which feels a bit messy. PSSA solves this problem by letting you run `invoke-scriptanalyzer` using a specific psd1 against different -path values.

It's the way PSES currently implements PSSA, using -ScriptDefinition, that limits us atm. So why introduced many files in a repo to solve a editor-specific issue - especially when the editor already uses settings.json for other PSSA settings?

Or this could possibly be delayed and fixed as part of the PSSA2-rewrite since better integration-support for editors is one of the suggested features for the major release. 🙂

**Comment by SydneyhSmith** [comment]
Thanks for the clarifications, one more question--are you seeing performance impacts or other issues with the open files that you would not like to be utilized, or is the issue more along the issue of not wanting to see squiggles/ or other problems? 
(would be great to better understand why you dont want the files to be analyzed)...thanks!

**Comment by fflaten** [comment]
Currently to avoid unnecessary warnings/squiggles. Haven't seen any performance impacts (yet). Our files are probably to small for that to happen. 

Can't speak for others though. I have seen a couple of earlier issues about excluding folders as well.

**Comment by dbaileyut** [comment]
The GitHub Copilot completions panel has a ton of errors detected when it's open and I would love to exclude those at the user settings level rather than per folder or workspace but still have PowerShell syntax highlighting.

![image](https://github.com/PowerShell/vscode-powershell/assets/7095040/65ebbb3f-1d03-4cd7-88f8-38990c0b99de)

**Comment by Pxtl** [comment]
I hit F12 on a call to an external referenced module (that I had put in a `/myproject/packages` dir) and it listed a zillion problems, so I'd appreciate the ability to have some kind of `.analyzerignore` folder concept.  Or have it just ignore the files I've already `.gitignore`d (since I've already `.gitignore`d the packages folder), but I appreciate that's kind of presumptuous.

Also I didn't realize it only runs on open files.  That's different from other tools, is there a toggle to change that?

**Comment by Pxtl** [comment]
> So I think adding a configuration like this to PSScriptAnalyzer's settings file makes sense, and for that I'd ask you to open a feature request issue in the PSScriptAnalyzer repo.

They said to open this issue here.

https://github.com/PowerShell/PSScriptAnalyzer/issues/561#issuecomment-552165652

**Comment by JustinGrote** [comment]
@pxtl it requires contributions in both for the ideal solution. [I've noted my thoughts in the PSScriptAnalyzer issue.](https://github.com/PowerShell/PSScriptAnalyzer/issues/561#issuecomment-2452189176)
